### PR TITLE
Add flag support and refactor code

### LINKED ISSUE #3: Cleanup the code into individual packages
Right now the whole of code is in the main.go, move it to individual packages for docker, containerd, fanotify, etc.

### LINKED ISSUE #5: Add support for flags
It would be easy to make the code more configurable. We could use https://github.com/spf13/cobra

### PR DESCRIPTION
**Pull request by knrt10**

### CODE REVIEW

**Comment by knrt10** [comment]
@surajssd, the code is building fine but not working. Looking into it. Can you please check if this is the correct layout for packages?

**Comment by surajssd** [comment]
@knrt10 can you post the errors?

File: cli/cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+func init() {
+	log.SetLevel(log.DebugLevel)
+	RootCmd.DisableAutoGenTag = true
+
+	pf := RootCmd.PersistentFlags()
+	pf.StringVarP(&hostname, "hostname", "", "", "Name of node in k8s control plane")
+	pf.StringVarP(&hostRuntime, "runtime", "", "docker", "Name of k8s container runtime")
+	pf.StringVarP(&kubeconfig, "kubeconfig", "", "~/.kube/config", "Path to kubeconfig")
```
Comment (surajssd) [inline cli/cmd/root.go]: How about `$HOME/.kube/config` that way we can use os.ExpandEnv?

File: internal/util.go [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+}
+
+const (
+	podKey   = "enforce.k8s.io"
+	podValue = "deny-third-party-execution"
+)
+
+const (
+	RuntimeDocker     = "docker"
+	runtimeContainerd = "containerd"
```
Comment (surajssd) [inline internal/util.go]: why is one exported and other isn't?

**Comment by knrt10** [comment]
@surajssd, fixed. Please review

File: cli/cmd/root.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	internal.CheckRuntime(hostRuntime)
+}
+
+func fanotify(hostname, hostRuntime, kubeconfig string) {
+	pods := make(map[string]*v1.Pod)
+	go internal.GetNewPods(pods, hostname, kubeconfig)
```
Comment (surajssd) [inline cli/cmd/root.go]: How about a k8s package? And this function lives there?

File: internal/util.go [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+)
+
+var containerdNamespace string
+
+func CheckRuntime(hostRuntime string) {
```
Comment (surajssd) [inline internal/util.go]: this function is a bit misleading. It checks runtime but also modified the containerd namespace.

File: pkg/containerd/containerd.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+)
+
+const (
+	ContainerdSocket = "/run/containerd/containerd.sock"
+)
+
+type Container struct {
+	*pb.ContainerDefinition
+	*oci.Spec
+}
```
Comment (surajssd) [inline pkg/containerd/containerd.go]: This is not containerd specific.

File: pkg/containerd/containerd.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+	return cnts[0], closer, nil
+
+}
+
+func GetContainer(cntIG *pb.ContainerDefinition, oci *oci.Spec) *Container {
+	return &Container{
+		cntIG, oci,
+	}
+}
```
Comment (surajssd) [inline pkg/containerd/containerd.go]: This is also not containerd specific.

File: internal/util.go [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+		podName = container.Labels["io.kubernetes.pod.name"]
+	default:
+		return "", fmt.Errorf("no kind label on container")
+	}
+
+	cntName := container.Labels["io.kubernetes.container.name"]
+	podNamespace := container.Labels["io.kubernetes.pod.namespace"]
+	podUID := container.Labels["io.kubernetes.pod.uid"]
+
+	return "k8s_" + podName + "_" + cntName + "_" + podNamespace + "_" + podUID, nil
```
Comment (surajssd) [inline internal/util.go]: this whole part can be moved to containerd package.

**Review by surajssd — CHANGES_REQUESTED** [review]

File: internal/util.go (hunk 2 above)
Comment (knrt10) [reply internal/util.go:42]: exported both, but the RuntimeDocker was also getting used for checking inside `root.go`.

**Comment by knrt10** [comment]
@surajssd, updated. PTAL

File: internal/util.go (hunk 2 above)
Comment (knrt10) [reply internal/util.go]: Fixed. Moved them to respective packages

File: README.md [hunk 8]
```diff
@@ -13,3 +13,50 @@ sudo ./fanotify-poc ROOTFS_PATH
...
+  - ### Containerd based cluster:
+      - Clone the Kubernetes repository.
+      - Once cloned inside the repo run the following commands:
+
+      ```bash
+      # This should be the IP of the machine like eth0 or whatever default NIC is.
+      export API_HOST=
+
+      ./hack/local-up-cluster.sh
+      ```
```
Comment (surajssd) [inline README.md]: Cluster creation is out of scope of this document.

File: cli/cmd/root.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+		os.Exit(1)
+	}
+}
+
+func init() {
+	log.SetLevel(log.DebugLevel)
+	RootCmd.DisableAutoGenTag = true
+
+	pf := RootCmd.PersistentFlags()
+	pf.StringVarP(&hostname, "hostname", "", "", "Name of node in k8s control plane")
```
Comment (surajssd) [inline cli/cmd/root.go]: It does not have to be control plane, it is just the name of the node on which this app is running.

**Review by surajssd — COMMENTED** [review]
Just some nits, otherwise looks good to go.

**Review by surajssd — APPROVED** [review]

**Pull request merged**